File: javascript/linkedliststack.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedListStack {
  constructor() {
    this.head = null;
  }

  // method to add element to stack
  push(data) {
    const newNode = new Node(data);
    newNode.next = this.head;
    this.head = newNode;
  }

  // method to remove and return top element of stack
  pop() {
    if(this.isEmpty()) {
      throw new Error("stack is empty");
    }
    const poppedNode = this.head;
    this.head = this.head.next;
    return poppedNode.data;
  }

  // method to return top element of stack without removing it
  peek() {
    if(this.isEmpty()) {
      throw new Error("stack is empty");
    }
    return this.head.data;
  }
  // method to check if stack is empty
  isEmpty() {
    return this.head === null;
  }
  // method to get size of stack
  size() {
    let count = 0;
    let current = this.head;
    while(current) {
      count++;
      current = current.next;
    }
    return count;
  }

  // method to clear stack
  clear() {
    this.head = null;
  }
}
// example usage
const stack = new LinkedListStack();
stack.push(10);
stack.push(20);
stack.push(30);

console.log("top element: ", stack.peek());
console.log("stack size: ", stack.size());
console.log("popped element:", stack.pop());
console.log("stack size after pop: ", stack.size());

File: c/linkedListStack.c
#include <stdio.h>
#include <stdlib.h>

// node structure for the stack
struct Node {
  int data;
  struct Node* next;
};

// function to create a new node
struct Node* newNode(int data) {
  struct Node* node = (struct Node*)malloc(sizeof(struct Node));
  node->data = data;
  node->next = NULL;
  return node;
}

// function to push element onto stack
void push(struct Node** head_ref, int data) {
  struct Node* new_node = newNode(data);
  new_node->next = (*head_ref);
  (*head_ref) = new_node;
}

// function to pop element from stack
int pop(struct Node** head_ref) {
  if(*head_ref == NULL) {
    printf("stack underflow\n");
    return -1;
  }
  struct Node* temp = *head_ref;
  int popped = temp->data;
  (*head_ref) = temp->next;
  free(temp);
  return popped;
}

// function to peek at top element of stack 
int peek(struct Node* head) {
  if(head == NULL) {
    printf("stack is empty\n");
    return -1;
  }
  return head->data;
}

int main() {
  struct Node* head = NULL;
  push(&head, 10);
  push(&head, 20);
  push(&head, 30);

  printf("popped element: %d\n", pop(&head));
  printf("top element: %d\n", peek(head));

  return 0;
}
